src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -887,41 +912,109 @@ public static unsafe void MulElementWiseU(Span<float> src1, Span<float> src2, Sp
+                        case 5: res += pValues[4]; goto case 4;
+                        case 4: res += pValues[3]; goto case 3;
+                        case 3: res += pValues[2]; goto case 2;
+                        case 2: res += pValues[1]; goto case 1;
+                        case 1: res += pValues[0]; break;
+                    }
+
+                    return res;
                 }
 
-                result256 = VectorSum256(in result256);
-                Vector128<float> resultPadded = Sse.AddScalar(Avx.GetLowerHalf(result256), GetHigh(result256));
+                Vector256<float> result = Avx.SetZeroVector256<float>();
 
-                Vector128<float> result128 = Sse.SetZeroVector128();
+                nuint address = (nuint)(pValues);
+                int misalignment = (int)(address % 32);
+                int remainder = 0;
 
-                if (pSrcCurrent + 4 <= pSrcEnd)
+                if ((misalignment & 3) != 0)
                 {
-                    result128 = Sse.Add(result128, Sse.LoadVector128(pSrcCurrent));
-                    pSrcCurrent += 4;
+                    // Handles cases where the data is not 32-bit aligned and we can't ever use aligned operations
+
+                    remainder = length % 8;
+
+                    for (float* pEnd = pValues + (length - remainder); pValues < pEnd; pValues += 8)
+                    {
+                        Vector256<float> temp = Avx.LoadVector256(pValues);
+                        result = Avx.Add(result, temp);
+                    }
                 }
+                else
+                {
+                    if (misalignment != 0)
+                    {
+                        // Handle cases where the data is not 256-bit aligned by doing an unaligned read and then
+                        // masking any elements that will be included in the first aligned read
 
-                result128 = SseIntrinsics.VectorSum128(in result128);
+                        misalignment >>= 2;
+                        misalignment = 8 - misalignment;
 
-                while (pSrcCurrent < pSrcEnd)
+                        Vector256<float> temp = Avx.LoadVector256(pValues);
+                        Vector256<float> mask = Avx.LoadVector256(((float*)(pLeadingAlignmentMask)) + (misalignment * 8));
+                        temp = Avx.And(temp, mask);
+                        result = Avx.Add(result, temp);
+
+                        pValues += misalignment;
+                        length -= misalignment;
+                    }
+
+                    if (length > 7)
+                    {
+                        // Handle all the 256-bit blocks that we can now that we have offset to an aligned address
+
+                        remainder = length % 8;
+
+                        for (float* pEnd = pValues + (length - remainder); pValues < pEnd; pValues += 8)
+                        {
+                            Vector256<float> temp = Avx.LoadAlignedVector256(pValues);
```

Comment:
Note: I need to change this to use `LoadVector256` for better codegen.

The address is aligned, which prevents reads that would cross a cach-line or page boundary (the important bit for perf, since newer hardware otherwise does unaligned reads that are as fast as aligned reads). However, the JIT will only fold unaligned-loads when the VEX-encoding is being used (in contrast to only folding aligned-loads when the legacy-encoding is being used). This is due to the different semantics for the memory operand on the newer encoding.